### PR TITLE
Feature/cinema pages

### PR DESCRIPTION
**Pull request by Hoookie**

### CODE REVIEW

File: src/app/cinema/components/hall-seats-configurator/hall-seats-configurator.component.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { Component, OnInit, forwardRef } from '@angular/core';
+import { animate, style, transition, trigger } from '@angular/animations';
+import { ValidateFormService } from 'src/app/shared/services/validate-form.service';
+
+@Component({
+  selector: 'app-hall-seats-configurator',
+  templateUrl: './hall-seats-configurator.component.html',
+  styleUrls: ['./hall-seats-configurator.component.scss'],
+  providers: [{
+    provide: NG_VALUE_ACCESSOR,
```
Comment (viento006) [inline src/app/cinema/components/hall-seats-configurator/hall-seats-configurator.component.ts]: Why is it a formcontrol, not a form?

File: src/app/cinema/guards/cinema-exists.guard.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+@Injectable()
+export class CinemaExistsGuard implements CanActivate {
+
+  constructor(private cinemaService: CinemaService) { }
+
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
+    const { cinemaId } = route.params;
+
+    return this.cinemaService.getById(cinemaId).pipe(
```
Comment (viento006) [inline src/app/cinema/guards/cinema-exists.guard.ts]: How many requests do you make while entering this page?

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.scss [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+
+  &__save-button {
+    position: absolute;
+    bottom: 20px;
+    right: 20px;
+    font-size: 1.1rem;
+  }
+}
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.scss]: Line

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+          [ngStyle]="{ width: defaultSeatWidth * seatData.count + 'px' }">
+            {{ seatData.seat }}
+          </div>
+        </div>
+      </div>
+      <div class="hall-builder__row-number">{{ rowNumber(i) }}</div>
+    </div>
+  </div>
+  <button (click)="save()" class="hall-builder__save-button form-group__button form-group__button--primary">Save</button>
+</div>
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.html]: LIne

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.html [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      [seatTypes]="seatTypes"></app-hall-rows-configurator>
+    <app-hall-seats-configurator formControlName="seatTypes"></app-hall-seats-configurator>
+  </form>
+  <div class="hall-builder__preview">
+    <div *ngFor="let row of rows; let i = index" class="hall-builder__row">
+      <div class="hall-builder__row-number">{{ rowNumber(i) }}</div>
+      <div class="hall-builder__row-groups">
+        <div *ngFor="let rowGroup of createRowGroups(row); let seatIndex = index" class="hall-builder__row-group">
+          <div *ngFor="let seatData of rowGroup.seats" class="hall-builder__row-seat"
+          [ngStyle]="{ width: defaultSeatWidth * seatData.count + 'px' }">
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.html]: Add a tab

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+  ngOnDestroy() {
+    this.unsubscribe$.next();
+    this.unsubscribe$.complete();
+  }
+
+  save() {
+    const errors = this.hallService.validateForm(this.hall);
+    if (errors.length) {
+      errors.forEach(error => this.alertService.danger(error, 'Error'));
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts]: let's throw 500 alerts in front of user!

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    let counter = 0;
+    for (const rowGroup of row) {
+
+      const rowSeatType = this.seatTypes.find(seatType => seatType.id === rowGroup.type);
+
+      if (rowGroup.count && rowGroup.type && rowSeatType) {
+        const seats = [];
+        const seatCount = rowSeatType.count;
+        Array(rowGroup.count).fill(0).forEach(() => {
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts]: for loop?

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  rowNumber(index: string) {
+    return parseInt(index, 10) + 1;
+  }
+
+  createRowGroups(row: IRowGroup[]) {
+    const rowGroups = [];
+
+    let counter = 0;
+    for (const rowGroup of row) {
+
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts]: Remove empty line

File: src/app/cinema/pages/create-cinema/create-cinema.component.ts [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        tap(cinema => {
+          this.titleService.setTitle(`Edit cinema: ${cinema.title}`);
+        }),
+        catchError(() => {
+          return this.router.navigateByUrl('/cinemas');
+        }),
+        takeUntil(this.unsubscribe$)
+      )
+      .subscribe((cinema: ICinema) => {
+        document.querySelector('.create-cinema-card').classList.remove('card--disabled');
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema/create-cinema.component.ts]: wtf is going on? where is jQuery?

File: src/app/cinema/pages/create-cinema/create-cinema.component.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        this.loadCinemaToForm(cinema);
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema/create-cinema.component.ts]: You can patch form just here, withoudt calling a method

File: src/app/cinema/pages/create-cinema/create-cinema.component.ts [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    });
+  }
+
+  loadHalls() {
+    this.hallService.get(this.cinemaId).subscribe(halls => {
+      this.halls = halls;
+    });
+  }
+
+  prevalidateForm() {
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema/create-cinema.component.ts]: i beleive it is submit form

File: src/app/cinema/pages/create-cinema/create-cinema.component.ts [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        filter((value: string) => !!value.length),
+        tap(title => service.title = title),
+        switchMapTo(
+          this.modal
+            .showModal('Create service', 'Service description:', true, { promptType: 'textarea' })
+        ),
+        tap(description => service.description = description),
+        switchMapTo(
+          this.modal
+            .showModal('Create service', 'Service price:', true)
```
Comment (viento006) [inline src/app/cinema/pages/create-cinema/create-cinema.component.ts]: Why not to use just one single modal instead?

File: src/app/shared/components/modal/modal.component.ts [hunk 13]
```diff
@@ -9,16 +9,30 @@ import { FormControl, Validators } from '@angular/forms';
 })
 export class ModalComponent {
   promptControl = new FormControl('');
-  isPromptValid = true;
   isShown = false;
+  isInput = false;
+  modal: any = {};
```
[the remark this answers is not in the document]
Comment (Hoookie) [reply src/app/shared/components/modal/modal.component.ts]: Will be in next pr

File: src/app/shared/services/modal.service.ts [hunk 14]
```diff
@@ -1,51 +1,61 @@
...
+import { BehaviorSubject, Subject, Observable } from 'rxjs';
+import { filter, debounceTime } from 'rxjs/operators';
+
+interface IPromptOptions {
+  promptType?: 'default' | 'textarea';
+  promptDefaultValue?: string;
+}
 
 @Injectable({
   providedIn: 'root'
```
[the remark this answers is not in the document]
Comment (Hoookie) [reply src/app/shared/services/modal.service.ts]: Will be, when i rework modal system. Also move all root services to forRoot method

File: src/app/shared/components/modal/modal.component.html [hunk 15]
```diff
@@ -1,9 +1,10 @@
-<div (click)="cancel()" class="overlay" [ngClass]="{'overlay--show': modal.state}"></div>
-<div class="modal" [ngClass]="{'modal--open': modal.state}">
+<div (click)="cancel()" class="overlay" [ngClass]="{'overlay--show': isShown}"></div>
+<div class="modal" [ngClass]="{'modal--open': isShown}">
   <div class="modal__header">{{ modal.title }}</div>
   <div class="modal__body">
     <p class="modal__text">{{ modal.text }}</p>
-    <input *ngIf="modal.isPrompt" class="modal__input">
+    <input *ngIf="modal.isPrompt && isInput" class="modal__input" [formControl]="promptControl">
+    <textarea *ngIf="modal.isPrompt && !isInput" class="modal__textarea" [formControl]="promptControl"></textarea>
```
[the remark this answers is not in the document]
Comment (Hoookie) [reply src/app/shared/components/modal/modal.component.html]: Next pr

File: src/app/shared/services/modal.service.ts [hunk 16]
```diff
@@ -1,51 +1,61 @@
...
-    this.modalText = text;
-    this.isModalPrompt = isPrompt;
-    this.isModalOpen.next(true);
-    return this.modalResponse.pipe(
-      take(1),
-      tap(() => this.isModalOpen.next(false)),
-      map((response: { value: string }) => {
-        if (!response) {
-          return new Error('Modal was closed.');
+  showModal(title: string, text = '', isPrompt = false, options?: IPromptOptions) {
```
[the remark this answers is not in the document]
Comment (Hoookie) [reply src/app/shared/services/modal.service.ts]: Next pr

File: src/app/cinema/services/hall.service.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    return errors;
+  }
+
+  private validateSeatTypes(seatTypes: ISeatType[]): string[] {
+    const errors = [];
+    if (!seatTypes.length) {
+      errors.push('You should provide minimum 1 seat type to hall.');
+    }
+    for (const seatType of seatTypes) {
```
[the remark this answers is not in the document]
Comment (Hoookie) [reply src/app/cinema/services/hall.service.ts]: Next pr

File: src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts (hunk 7 above)
Comment (Hoookie) [reply src/app/cinema/pages/create-cinema-hall/create-cinema-hall.component.ts]: Next pr

**Pull request merged**